### ISSUE TITLE
Next feature release

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#8827]
Knowing, that we just have done a feature release I do see a lot of activity with code changes and corresponding new features (which I really appreciate).
In order to get early user feedback and not let the amount of changes (and thus possible new problems) grow too much, I would like to do another branch merge and deploying a beta version rather soon.

I do however need your feedback about when we reached a state where x new features and changes + their fixes and addons are done on `master`.

**Comment by moving-bits** [comment]
If we want to release rather soon we should collect issues we want to include, and issues we should postpone to a later version. I've included one example per category - feel free to add or move items around:

## Issues to be fixed before creating a beta version
- #8821 Map file download fails if target dir not writable

## Issues to be fixed until release (but not necessarily before first beta)
- #8818 Significant amount of users report login problem "Username/Password incorrect"

## Issues to be postponed to a later version
- #8822 [Related to Issue:#2553]: store offline log completely - DataStore part

**Comment by eddiemuc** [comment]
In case of #8822 postphoned I will convert it to draft and build GUI/logic changes on top of it for now

**Comment by Lineflyer** [comment]
I think the proposal of @moving-bits sounds good as #8822 is a bigger upcoming change, which might need more time.

So lets wait for #8821 and (maybe) #8818

**Comment by fm-sys** [comment]
#8793 should also be fixed until release IMHO couse it is kind of regression

**Comment by eddiemuc** [comment]
Please add #8834, this is a regression bug (which unfortunately I introduced myself before)

**Comment by moving-bits** [comment]
@fm-sys, @eddiemuc 
I've added both (and more) issues to the list above, also mentioning PR numbers provided as fixes, if available.

**Comment by fm-sys** [comment]
I've coded #8836 now. The pull request will follow this afternoon. So please wait for it before merging branches...

**Comment by fm-sys** [comment]
> 
> 
> I've coded #8836 now. The pull request will follow this afternoon. So please wait for it before merging branches...

See PR #8881

**Comment by bekuno** [comment]
It would also be fine if these bugs are fixed before releasing, imho.
 - #7429 (Duplicate between projected WP and personal note WP not detected)
 - #8314 (Rounding Error at Waypoint from Cache Coordinates)
 - #8845 (Check coordinate parsing for the format used by Locus Map)

**Comment by Lineflyer** [comment]
IMHO we should slow down a bit.. I think we should wait for #8881 for now and then start a new beta.
Those bugs @fm-sys mentioned, are no immediate regressions AFAIK.

Once we merged branches however the work can continue o fix e.g. these bugs.

**Comment by moving-bits** [comment]
AFAICS we are ready to prepare a beta version now, having all issues in [this list](https://github.com/cgeo/cgeo/issues/8827#issuecomment-675096424) fixed (and many more).

Crowdin-updates and maybe #8899 can be merged later, as well as a fix for #8893 (where we need more input for first).
For #8818 we have some extended logging since #8867 - let's see if that does help us to get a grip on it.

**Comment by Lineflyer** [comment]
Sounds good. 

- #8899 can IMHO be merged after change (see my comment there).

- We should also wait for crowdin french to be fit as the updates target master (OK, we could also branch again from master later on, but only if no functional changes are applied until then).

**Comment by Lineflyer** [comment]
Due to my personal schedule (I will be on vacation 05.09.-21.09), I would be keen (but its no must) to get out a new beta during this week (before 05.09.).

We can merge the Crowdin translations as last step before branching to take all the corrections incoming until then.

Anything else ongoing, which should be included?

**Comment by moving-bits** [comment]
> Anything else ongoing, which should be included?

I'm still trying to get a grip on #8893, which happens only on some devices.
Other than that nothing from my side for the upcoming release.

**Comment by moving-bits** [comment]
Thanks to @fm-sys's testing efforts we have a fix for #8893 finally. Changelog is updated and Crowdin is merged.
=> ready to go... @Lineflyer

**Comment by Lineflyer** [comment]
OK, will wait for some Crowdin udpates regarding ChripWolf and merge branches during this week.
Please wait with further merges not related.

**Comment by Lineflyer** [comment]
Add on:
I have just sent out a message to Crowdin users for a "final round" before next release.

**Comment by Lomanic** [comment]
No love for #987? While I am on vacation for the remaining week, I would have also appreciated a new version (even beta) including it. Because I saw some issues in nightlies recently, I probably won't use nightlies for this feature as I wrote there (I'm not even sure I have sufficient space to make a cgeo backup to rollback from nightlies also).

**Comment by Lineflyer** [comment]
Ready for merging `master` to `release`?
I am currently importing from Crowdin. After that I will merge branches if no one stops me.

**Comment by moving-bits** [comment]
I'm not gonna stop you ;-)

**Comment by Lineflyer** [comment]
Attention!
`release` has just been forwarded to HEAD of `master` to initiate the beta phase.

This means:
- Important bugfixes for next release should be merged to `release` and then `release` merged upstream into `master`
- New things can now be merged on `master` for the next feature release round

Will smoke test and publish new beta soon.

**Comment by Lineflyer** [comment]
@moving-bits 
Looking at the change log:
Might it be, that it is a little to detailed and redundant. As we should write it targeted to the normal release user (incl. beta) we should focus on changes between releases rather than improvements on features not yet released before.
This came to my mind, when reading about the individual track: Isn't this a new feature for the upcoming release? In this case it should rather be a single change, or not?

**Comment by moving-bits** [comment]
I've tried to include only things relevant to users of current release, not changes from nightly to nightly, but I may have missed one.

As for the GPX track: This is already available in current release, being named "Individual track" there. (Not to be confused with "Individual route" - that's why we renamed it after the last release.)

**Comment by Lineflyer** [comment]
Beta version 2020.09.02-RC is on its way...at least I hope so as there is a failure message on Google Play:
I made the failure to not deactivate 2020.07.24-RC-legacy for that beta, now Google Play is claimed a version conflict "blocked by production". From logical point of view this should only affect legacy users (API 16-20) who do not participate in that beta (API 21+) anyhow. Still I am not sure about Google Play logic at this point. 

Lets just wait and see if the beta rolls out.

**Comment by Lineflyer** [comment]
> that's why we renamed it after the last release.

Thanks for clarification. This explains my confusion as well.

**Comment by Lineflyer** [comment]
According to status.cgeo.org the beta seems to reach our users. So I think my logic was right and the beta rollout is working.

**Comment by Lineflyer** [comment]
I briefly checked the Google Play console and opened an issue for a new crash which looks like a regression.

As announced I will be on vacation starting tomorrow. So if anyone who has access to Google Play console can check there from time to time, it would be nice.

**Comment by moving-bits** [comment]
I'll try to keep an eye on the console messages.
Have a nice vacation!

**Comment by geocachermgo** [comment]
When will be the next release? What is the normal delay between beta and release?

**Comment by moving-bits** [comment]
> When will be the next release? What is the normal delay between beta and release?

A typical beta cycle is two or three weeks, depending on available time of the team members.
Normally there's only one beta, but sometimes we have two or even more beta versions, depending on how smooth the beta behaves throughout the beta cycle.

**Comment by moving-bits** [comment]
How to continue in current release cycle?

Looking at the Play Console logs, there are mainly the `java.lang.ClassCastException` and `java.lang.RuntimeException in android.os.Parcel.readValue` marked as new for beta version, both of which #8957 tries to address. As there are also a couple of more fixes on `release` branch I propose to publish a second beta version to see if those fixes work. If everything's well, a one week cycle may be enough for second beta, IMHO, depending on support's inbox.

@Lineflyer AFAIK you're the only one currently able to publish a beta. Do you have time and technical access to do so, or when will be a good spot?

**Comment by geocachermgo** [comment]
I would like to have the new attributes included.

**Comment by moving-bits** [comment]
> I would like to have the new attributes included.

This PR has been prepared for and merged to `master` branch, thus currently would not be available in the next release.

**Comment by moving-bits** [comment]
@Lineflyer 
We should prepare a second beta anyway, as there have been a couple of changes on `release`, and we need to test if the fix for the two errors in Play Console mentioned above do work.
How about doing another merge from `master` to `release` (which would, as a side effect, include the new icons)? Due to vacation time the larger changes for offline logging etc. are a few weeks in the nightlies already. 

What do you think? Do another merge from `master` to `release` or stay with current `release`? How stable is `master` from a support perspective?

**Comment by Lineflyer** [comment]
@moving-bits I dont think it would be wise to merger from master again. Too many changes and quite some regressions AFAICS from my mails (e.g. the two line layout).
We can however push a second beta for some days, then release and then merge branches again for another quick feature release cycle (another beta).

I will return home tomorrow and try to get an overview. Quite a lot of activity and I was not able to follow all the hundreds of Github mails I received :)

**Comment by moving-bits** [comment]
> @moving-bits I dont think it would be wise to merger from master again. Too many changes and quite some regressions AFAICS from my mails (e.g. the two line layout).
> We can however push a second beta for some days, then release and then merge branches again for another quick feature release cycle (another beta).

Ok, understood & agreed. (Although "two line layout" should be resolved in most aspects.)

> I will return home tomorrow and try to get an overview. Quite a lot of activity and I was not able to follow all the hundreds of Github mails I received :)

Sorry, I thought you were already back from vacation.

**Comment by Lineflyer** [comment]
Back from vacation and briefly looked over the changes on both branches:

I have to admit, that it would indeed make sense to merge from `master` again, however IMHO the following issues should be adressed before doing this:
- [ ] #9013 -> Improve +x explanation text (PR is ready to merge, isn't it?)
- [ ] #8970 -> Meaning of image prefix (will comment into that issue)
- [ ] #9018 -> String fix for log image (use plural)

Anything else?

If we decide to go this way, please wait with adding unrelated changes until branches are merged.

**Comment by eddiemuc** [comment]
With regards to #8970 : If you decide to merge `release` from `master` again, then I would for now remove the "image prefix" text field completely from GUI (and implement a more sustainable solution with #8978 for next release)

**Comment by moving-bits** [comment]
For #8818 we need someone having those login problems install #9025 APK first to get an extended log for further investigation. - Currently we don't have a clue where those problems come from.

**Comment by Lineflyer** [comment]
> For #8818 we need someone having those login problems install #9025 APK first to get an extended log for further investigation. - Currently we don't have a clue where those problems come from.

Yes and no. We just need a better way to identify a credential issue in the first step. Currently all(!) users with login problems report they receive a "username/password error" and thus I am unable to distinguish which user on support has our "unknown" problem and which user only entered the wrong credentials. :(

**Comment by moving-bits** [comment]
If I understand the source correctly, we currently either get a "Wrong username and/or password" message or an "Unknown error", and in a different place we have a string "Login failed".

The first text only gets send if the HTML returned by the server contains the string "Your password or username/email is incorrect". What I can check, though, is whether this string is always part of the source as standard text variable, even if we do not have the specific error situation ("false positive").

**Comment by Lineflyer** [comment]
I am pretty sure we have false positives, as there is NO report of "unknown" error but really many reports of incorrect credentials, where users confirm that they already doublechecked. 
So the last change in this area made the detection of unknown errors fail and always return credential failure to the user.

**Comment by moving-bits** [comment]
The last two changes I remember are adding a `Log.d(...)` statement (with #8867) and removing it again (with #9003), which should not change the behavior in this case. Before that we changed the string to the one currently used (with #8530).

**Comment by eddiemuc** [comment]
For #8970 , a PR is provided: #9034

**Comment by bekuno** [comment]
PR #9044 to fix #9038 is ready.
Because this is a critical error we should go the way to the next beta release and let the beautifying things for the following release.

**Comment by moving-bits** [comment]
#9044 is merged. For the upcoming second beta phase, based on a new merge from master to release, I would say:

- include #9054, #9043
- maybe include #9055 (does not harm, but would help in support work)
- do not include the other open PR

We do not have any clues on #8818 yet, so we should not wait for it now (but need to continue to work on it).

Anything else which should be fixed before next beta?

**Comment by Lineflyer** [comment]
> include #9054, #9043

OK for me. First one already merged. Feel free to merge #9043 if OK codewise.

Regarding #8818 and #9056 it would be fantastic if we (in a first step) just manage to distinguish the different error reasons and provide proper feedback to users.
Minimum approach would be to fix the current problem, that all errors show "credentials wrong". The best approach would be to also detect the captcha problem. 
I guess finding a solution for the captcha problem - if any possible - needs more time and trying and should not be included now into the next beta/release. Please be reminded, that changing login procedures is "dangerous" as it is a critical function.

**Comment by moving-bits** [comment]
Regarding #8818:
- #9039 is merged - should be able to distinguish between "real" login errors and false positives
- #9025 (not to be merged) is an updated test APK - should be able to detect captcha errors

**Comment by eddiemuc** [comment]
In case master is merged to release, please include #9051 / #9066 (compass view) as well, because it fixes a problem which was introduced with current master as well

**Comment by Lineflyer** [comment]
> In case master is merged to release, please include #9051 / #9066 (compass view) as well, because it fixes a problem which was introduced with current master

Done!

So ready for merge now? IMHO we are.

**Comment by moving-bits** [comment]
> > In case master is merged to release, please include #9051 / #9066 (compass view) as well, because it fixes a problem which was introduced with current master
> 
> Done!
> 
> So ready for merge now? IMHO we are.

I've added #9071 by @fm-sys (a one-liner for making `geo:` intents available for Firefox & Co).
Will do a last check for changelog, and then we're ready to go IMHO.

**Comment by Lineflyer** [comment]
Thanks @moving-bits 

`master` has been merged to `release` and changelogs migrated accordingly.

Will continue tonight or tomorrow with releasing a beta version from `release`

**Comment by fm-sys** [comment]
We of course will not have the shortened compass strings (#9074) in the beta, but I think it's even better this way. Otherwise there will be no translation for most languages, so we shouldn't merge them to `release` - the longer strings are probably fine for 99.9% of our users...

**Comment by Lineflyer** [comment]
Beta version 2020.09.26-RC has been uploaded to Google Play.

**Comment by Lineflyer** [comment]
> but I think it's even better this way. 

That was also my thinking when banching before merging #9074 
As soon as time permits I will take a closer look into branch management on Crowdin to overcome this problem in future.

**Comment by moving-bits** [comment]
> Beta version 2020.09.26-RC has been uploaded to Google Play.

Thanks @Lineflyer for preparing the beta, and thanks to all contributors to that version. Quite a few fixes, changes and new features in this version...

I'm looking forward to the results on Play Console, let's see how this version behaves.

**Comment by Lineflyer** [comment]
I have seen some issues filed by @moving-bits regarding Play Console reports. 
Anything critical things or easy to fix things still open?

From support perspective everything is rather calm, no negative user feedback for the beta version.
Shall we release?

**Comment by moving-bits** [comment]
AFAIC it's relatively calm on Play Console for the two beta versions. And if I'm not mistaken, there's only one error marked as new in the beta versions, and for that we have a fix/workaround.

Thus from my side: ready to go.

**Comment by Lineflyer** [comment]
What about #9120. 
This issue is asked quite a lot on support. Any objection to cherrypick it to `release` before releasing?

**Comment by Lineflyer** [comment]
I would prepare release tonight and if no objection arises (see above) I will include #9120 as well. Not sure though if it has any potential for problems?!

**Comment by Lineflyer** [comment]
I cherrypickes the commit from #9120 and build and smoke test release now.

BTW:
I will also push a legacy version as we have changes regarding the login failure reporting there.

**Comment by fm-sys** [comment]
> I will include #9120 as well. Not sure though if it has any potential for problems?!

Codewise there aren't any problems aspected. And from my personal tests it works like a charm... ;-)

**Comment by Lineflyer** [comment]
Status of release for both normal and legacy:

- [x] Check/Insert correct version name as headline in ```changelog-release.xml``` 
- [x] Build release on CI server
- [x] Perform smoke test with resulting ```cgeo-release.apk``` before continuing
- [x] Upload and publish ```cgeo-release.apk``` to Google Play
- [x] Tag the release on Github
- [x] Publish the release on Github including upload of ```cgeo-release.apk``` to Github
- [ ] Publish release on F-Droid 
- [ ] Trigger notification to status.cgeo.org via CI server once the release is available on Google Play
- [ ] Post release info to Facebook 
- [ ] Post release info to Twitter

**Comment by Lineflyer** [comment]
Yeah! 
New release 2020.10.04 and 2020.10.04-legacy are live.

Thanks to the team for your contributions and efforts!!!

BTW:
We are already 36 commits ahead on `master`, so next release seems not so far away 👍
